File: dataStructures/linked_lists/python/implementation.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

    def __last(self):
        n = self
        while n.next is not None:
            n = n.next
        return n

    def peek(self):
        return self.last().data


    def length(self):
        n = self
        counter = 0
        while n is not None:
            counter += 1
            n = n.next
        return counter

    def push(self, data):
        self.__last().next = Node(data)

    def pop(self):
        if self.next is not None:
            n = self
            while n.next.next is not None:
                n = n.next
            n.next = None

    def printList(self, separator = ", "):
        n = self
        print(n.data, end="")
        while n.next is not None:
            print(separator, end="")
            n = n.next
            print(n.data, end="")
        print("")

File: dataStructures/linked_lists/c++/implementation.cpp
#include<iostream>
#include<string>

template<class T>
class Node {  

    private:
    Node<T> *tail() {
        Node<T> *cur = this;
        while(cur->Next) {
            cur = cur->Next;
        }
        return cur;
    }
    public:
    T Data;
    Node<T>* Next;

    Node<T>(T value) {
        Data = value;
        Next = NULL;
    }
    
    T peek() const {
        return tail()->Data;
    }
    
    int length() const {
        Node<T> const *cur = this;
        int counter = 0;
        while(cur) {
            counter ++;
            cur = cur->Next;
        }
        return counter;
    }
    
    void push(T value) {
        tail()->Next = new Node<T>(value);
    }

    T pop() {
        Node<T> *cur = this;
        T last;
        while(cur->Next) {
            if(cur->Next->Next) {    
                cur = cur->Next;
            } else {
                last = cur->Next->Data;
                cur->Next = NULL;
            }
        }

        return last;
    }

};

int main(){
    std::string s = "";
    Node<int> a = Node<int>(4);
    a.push(3);
    std::cout << a.length();
    a.push(3);
    std::cout << a.length();
    a.pop();
    std::cout << a.length();
    std::cin >> s;
}
